packages/contracts-bedrock/src/cannon/interfaces/IPreimageOracle.sol

```diff
@@ -1,41 +1,89 @@
 // SPDX-License-Identifier: MIT
-pragma solidity 0.8.15;
+pragma solidity ^0.8.0;
+
+import { ISemver } from "src/universal/ISemver.sol";
+import { LPPMetaData } from "src/cannon/libraries/CannonTypes.sol";
+import { LibKeccak } from "@lib-keccak/LibKeccak.sol";
 
 /// @title IPreimageOracle
-/// @notice Interface for a preimage oracle.
-interface IPreimageOracle {
-    /// @notice Returns the length of the large preimage proposal challenge period.
-    /// @return challengePeriod_ The length of the challenge period in seconds.
-    function challengePeriod() external view returns (uint256 challengePeriod_);
+/// @notice Interface for the PreimageOracle contract.
+interface IPreimageOracle is ISemver {
+    /// @notice A raw leaf of the large preimage proposal merkle tree.
+    struct Leaf {
+        /// @notice The input absorbed for the block, exactly 136 bytes.
+        bytes input;
+        /// @notice The index of the block in the absorption process.
+        uint256 index;
+        /// @notice The hash of the internal state after absorbing the input.
+        bytes32 stateCommitment;
+    }
 
-    /// @notice Reads a preimage from the oracle.
-    /// @param _key The key of the preimage to read.
-    /// @param _offset The offset of the preimage to read.
-    /// @return dat_ The preimage data.
-    /// @return datLen_ The length of the preimage data.
-    function readPreimage(bytes32 _key, uint256 _offset) external view returns (bytes32 dat_, uint256 datLen_);
+    /// @notice Unpacked keys for large preimage proposals.
+    struct LargePreimageProposalKeys {
+        /// @notice The claimant of the large preimage proposal.
+        address claimant;
+        /// @notice The UUID of the large preimage proposal.
+        uint256 uuid;
+    }
+
+    error ActiveProposal();
```

Comment:
Looks like a lot of these errors duplicate what's in `CannonErrors.sol`. Should import those instead of redeclaring them?

Reply:
Hrmmm will check if this works

Reply:
You can see in PreimageOracle we do `import "src/cannon/libraries/CannonErrors.sol";`, but we typically prefer named imports. That will be noisier if there's a lot of errors being imported but I'm ok with that for the explicitness